### PR TITLE
refactor: remote catalog uses memory

### PR DESCRIPTION
**Pull request by MichaelScofield**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

This PR makes RemoteCatalogManager in datanode use MemoryCatalogManager. Should greatly reduce the counts of kvbackend requests.

This PR also adds `TableMetadataManager` to the structs it is expected to be used in the future.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
#1728

### CODE REVIEW

File: src/catalog/src/remote/manager.rs [hunk 1]
```diff
@@ -275,19 +211,7 @@ impl RemoteCatalogManager {
...
-            .max()
-        {
-            max_table_id = max_table_id.max(found_max_table_id);
-            info!(
-                "Catalog name: {}, max table id allocated: {}",
-                catalog_name, max_table_id
-            );
-        }
-
+        let _ = futures::future::try_join_all(joins).await?;
```
Comment (v0y4g3r) [inline src/catalog/src/remote/manager.rs:214]: Seems we don't have to explicitly ignore results now.

File: src/catalog/src/remote/manager.rs [hunk 2]
```diff
@@ -573,51 +438,42 @@ async fn open_or_create_table(
...
+    };
+    let registered = memory_catalog_manager.register_table(request).await?;
+    ensure!(
+        registered,
+        TableExistsSnafu {
+            table: table_ident.to_string(),
         }
-    }
+    );
+    info!("Successfully registered table, {table_ident}");
```
Comment (v0y4g3r) [inline src/catalog/src/remote/manager.rs:475]: nit: can we merge open/register logs?

File: src/catalog/src/remote/manager.rs [hunk 3]
```diff
@@ -573,51 +438,42 @@ async fn open_or_create_table(
...
+    info!("Successfully opened table, {table_ident}");
+
+    let request = RegisterTableRequest {
+        catalog: table_ident.catalog.clone(),
+        schema: table_ident.schema.clone(),
+        table_name: table_ident.table.clone(),
+        table_id,
+        table,
+    };
+    let registered = memory_catalog_manager.register_table(request).await?;
```
Comment (v0y4g3r) [inline src/catalog/src/remote/manager.rs:468]:
```suggestion
    let registered = memory_catalog_manager.register_table_sync(request)?;
```

File: src/catalog/src/remote/manager.rs [hunk 4]
```diff
@@ -638,92 +494,70 @@ impl CatalogManager for RemoteCatalogManager {
...
 
-        let _ = self
-            .register_table(
-                catalog_name.clone(),
-                schema_name.clone(),
-                request.table_name,
-                request.table.clone(),
-            )
-            .await?;
+        let registered = self.memory_catalog_manager.register_table(request).await?;
```
Comment (v0y4g3r) [inline src/catalog/src/remote/manager.rs:502]:
```suggestion
        let registered = self.memory_catalog_manager.register_table_sync(request)?;
```

File: src/common/meta/src/error.rs [hunk 5]
```diff
@@ -70,7 +70,7 @@ pub enum Error {
     #[snafu(display("Get null from cache, key: {}", key))]
     CacheNotGet { key: String, location: Location },
 
-    #[snafu(display("Failed to request MetaSrv, source: {}", source))]
+    #[snafu(display("{source}"))]
```
Comment (v0y4g3r) [inline src/common/meta/src/error.rs:73]: Why remove the error message?

File: src/meta-srv/src/service/store/kv.rs [hunk 6]
```diff
@@ -24,3 +36,109 @@ pub type ResettableKvStoreRef = Arc<dyn ResettableKvStore>;
 pub trait ResettableKvStore: KvBackend<Error = Error> {
     fn reset(&self);
 }
+
+/// An adaptor to bridge [KvStoreRef] and [KvBackendRef].
+pub struct KvBackendAdaptor(KvStoreRef);
```
Comment (v0y4g3r) [inline src/meta-srv/src/service/store/kv.rs:41]: Prefer `adapter` over `adaptor` in that:
```bash
$ find  .  -type f  -iname "*.rs" | xargs -L1 -I{} ggrep -Eo '[aA]dapter' {} | wc -l 
269

$ find  .  -type f  -iname "*.rs" | xargs -L1 -I{} ggrep -Eo '[aA]daptor' {} | wc -l 
66
```

File: src/catalog/src/remote/manager.rs (hunk 4 above)
Comment (MichaelScofield) [reply src/catalog/src/remote/manager.rs:502]: same above

File: src/common/meta/src/error.rs (hunk 5 above)
Comment (MichaelScofield) [reply src/common/meta/src/error.rs:73]: I'd like to display the error message that solely belong to the embedded BoxedError. At last, it's used in an adapter.

**Comment by MichaelScofield** [comment]
@v0y4g3r PTAL

**Comment by MichaelScofield** [comment]
@v0y4g3r PTAL

**Review by fengjiachun — APPROVED** [review]

**Review by v0y4g3r — APPROVED** [review]
LGTM

**Pull request merged**